### ISSUE TITLE
Training a 4-bit quantized model directly without Peft when model is loaded with  torch_dtype=torch.bfloat16

### ISSUE THREAD

**Issue by alexcpn** [huggingface/peft#1342]
### System Info

As per the documentation https://github.com/huggingface/blog/blob/main/4bit-transformers-bitsandbytes.md#can-we-train-4bit8bit-models it is not possible

```
Can we train 4bit/8bit models?

It is not possible to perform pure 4bit training on these models. However, you can train these models by leveraging parameter-efficient fine-tuning methods (PEFT) and train for example adapters on top of them. That is what is done in the paper and is officially supported by the PEFT library from Hugging Face.
```

However, with GPUs that support  `bfloat16` (A100 GPU), I was able to do direct training without LoRA adapters and with just BitsandBytes config by loading the model in `bfloat16`  and with BitsandBytes configuration of 4 bits

This is without any Instruction Training dataset generation directly on a raw generated text and the model training is susccessfull in that it is able to get the  knowledge of the  train text on cursory evaluation. 

See collab/git code - https://github.com/alexcpn/transformer_learn/blob/main/Mistral_7b_4bq_train.ipynb

colab -  https://colab.research.google.com/drive/1VDndzHRMBTBJ9vRVGML4F2Gf1KRQmnDl?usp=sharing

Code snippet below

```
# Load the entire model on the GPU 0
device_map = {"": 0} # lets load on the next
device = torch.device('cuda:0')

# Load base model
if bf16:
    torch_dtype=torch.bfloat16
else:
    torch_dtype=torch.float16

log.info(f"Going to load the model {model_name_long} in 4 bit Quanitsed mode {bnb_config} ")
# This works, this is training the qunatised model
model = AutoModelForCausalLM.from_pretrained(
    model_name_long,
    torch_dtype=torch_dtype,
    quantization_config=bnb_config,
    device_map=device_map
)

log.info(f"Loaded model in 4 bit Quantised form torch_dtype={torch_dtype}")
```

Log output, showing configuration

```
12:32:56 INFO:Going to load the model mistralai/Mistral-7B-Instruct-v0.1 in 4 bit Quanitsed mode BitsAndBytesConfig {
  "bnb_4bit_compute_dtype": "float16",
  "bnb_4bit_quant_type": "nf4",
  "bnb_4bit_use_double_quant": false,
  "llm_int8_enable_fp32_cpu_offload": false,
  "llm_int8_has_fp16_weight": false,
  "llm_int8_skip_modules": null,
  "llm_int8_threshold": 6.0,
  "load_in_4bit": true,
  "load_in_8bit": false,
  "quant_method": "bitsandbytes"
}
 
12:33:14 INFO:Loaded model in 4 bit Quantised form torch_dtype=torch.bfloat16
```
Training snippet

```
block_size = 250 # for uses 24/40 gb with model loaded A100 3 bad results
train_batch_size = 10 # 38 GB/40 in A100
if len_train_data > tokenizer.model_max_length:
    block_size = int(tokenizer.model_max_length/8) # tokenizer.model_max_length=1024
num_train_epochs = 50
save_epochs = 20

# Set the optimizer and learning rate scheduler
# num_warmup_steps = 100
max_grad_norm = 1.0
#optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4)
num_train_steps = len_train_data // train_batch_size * num_train_epochs
#lr_scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_train_steps)
log.info(f"len_train_data={len_train_data} block_size ={block_size} batch_size= {train_batch_size}")
model.train()

with autocast(dtype=torch.bfloat16):
    for epoch in range(num_train_epochs):
        log.info(f"Epoch {epoch+1} of {num_train_epochs}")
        epoch_loss = 0
        for i in range(0,len_train_data, block_size):
            # Get data in random per batch from input
            # not all training data may not be covered in one epoch here
            x,y= get_random_batch(len_train_data,input_ids,attention_mask,
                block_size=block_size,batch_size=train_batch_size)
            # attention_mask given by tokenize is array of ones= [1,1,..], that is attend to all tokens
            outputs = model(input_ids=x.to(device),attention_mask=y.to(device),labels=x.to(device))
            loss = outputs.loss
            epoch_loss += loss.item()
            loss.backward()
            #torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)
            optimizer.step()
            #lr_scheduler.step()
            optimizer.zero_grad()
```

Training loss 

```
01:13:13 INFO:Epoch 0 complete. Loss: 0.5360635197162629
...
01:24:39 INFO:Epoch 48 complete. Loss: 0.021451411843299864
01:24:39 INFO:Epoch 50 of 50
01:24:52 INFO:Epoch 49 complete. Loss: 0.02132726423442364
```
Ouput

training data - https://gist.githubusercontent.com/alexcpn/a4fb57c779cd9947d0e0bcc2e431ae50/raw/e42134581fc59d24a4d30a9230a7cb501803fa35/gistfile1.txt

```
 [INST]Who was Visgar and what was his proposition?[/INST] Visgar was a mysterious sorcerer from Jarkell's past. He approached Elorna with a proposition. He spoke of a potion that could grant its drinker visions of the future, a glimpse into what lay ahead. But for it to work, it needed an ingredient found only in Pentiagon.
Elorna, with her inherent curiosity, was tempted. But Visgar's condition was that she tell no one, not even Igodo.
However, unknown to Elorna, Visgar had other intentions. His true intent was to use the potion to manipulate the events of the past, to ensure the victory of Jarkell and the downfall of Pentiagon.
Elorna, after days of contemplation, decided to tell Igodo. Together, they formulated a plan to expose Visgar and save the kingdoms from the chaos caused by the potion.
The tale of Visgar and his proposition became a stark reminder of the fragility of balance. It strengthened the resolve of Elorna and Igodo to tread cautiously in their quest
````

Note - Training for smaller data sets like this, I used to create Instruction training data sets with another model and train. Direct training like this for bigger models on the small dataset using masked language modelling/ next word prediction based training, never used to work effectively.

Since this is working , why is the documentation telling that we cannot train 4bit/8bit loaded models directly

@TimDettmers 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

See git code - https://github.com/alexcpn/transformer_learn/blob/main/Mistral_7b_4bq_train.ipynb

colab - https://colab.research.google.com/drive/1VDndzHRMBTBJ9vRVGML4F2Gf1KRQmnDl?usp=sharing

### Expected behavior

As per the documentation training 4bit/8bit quantised models are not possible

**Comment by BenjaminBossan** [comment]
Yes, you are correct that it is technically possible to train a model quantized with bnb. When you look at the trainable parameters, you'll see that those are only `model.embed_tokens.weight`, `model.norm.weight`, `lm_head.weight`, and the `LayerNorm` layers. The point of using QLoRA is to train the attention layers (the query, key, and value weights), which is not possible without QLoRA (or similar techniques). This is what is meant by the quoted sentence.

Now whether it would be sufficient to train the mentioned layers without QLoRA, that's a different question. If you find that you can attain the same scores by only training those layers, that would certainly be an interesting finding.

**Comment by alexcpn** [comment]
You are right, only the layer normalisation layers are trainable along with lm_head

>If you find that you can attain the same scores by only training those layers, that would certainly be an interesting finding.

Maybe it worked because the data was very small. Still it is kind of very strange that it worked at all knowing this now

```
# Accessing layers and weights
for name, module in model.named_modules():
    print(f"Layer: {name}")
    for param_name, param in module.named_parameters():
        print(f"\tWeight: {param_name} | Size: {param.size()} | Trainable: {param.requires_grad}")  # Print first 2 values of each weight for brevity
````
Output concataneted
```
Weight: input_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
Weight: embed_tokens.weight | Size: torch.Size([32000, 4096]) | Trainable: True
Layer: model.norm
	Weight: weight | Size: torch.Size([4096]) | Trainable: True
Layer: lm_head
	Weight: weight | Size: torch.Size([32000, 4096]) | Trainable: True
```

```
Layer: 
	Weight: model.embed_tokens.weight | Size: torch.Size([32000, 4096]) | Trainable: True
	Weight: model.layers.0.self_attn.q_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: model.layers.0.self_attn.k_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: model.layers.0.self_attn.v_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: model.layers.0.self_attn.o_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: model.layers.0.mlp.gate_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: model.layers.0.mlp.up_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: model.layers.0.mlp.down_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: model.layers.0.input_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
	Weight: model.layers.0.post_attention_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
	....
	Weight: 31.post_attention_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
Layer: model.layers.0
	Weight: self_attn.q_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: self_attn.k_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: self_attn.v_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: self_attn.o_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: mlp.gate_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: mlp.up_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: mlp.down_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: input_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
	Weight: post_attention_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
Layer: model.layers.0.self_attn
	Weight: q_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: k_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: v_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: o_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
Layer: model.layers.0.self_attn.q_proj
	Weight: weight | Size: torch.Size([8388608, 1]) | Trainable: False
Layer: model.layers.0.self_attn.k_proj
	Weight: weight | Size: torch.Size([2097152, 1]) | Trainable: False
Layer: model.layers.0.self_attn.v_proj
	Weight: weight | Size: torch.Size([2097152, 1]) | Trainable: False
Layer: model.layers.0.self_attn.o_proj
	Weight: weight | Size: torch.Size([8388608, 1]) | Trainable: False
Layer: model.layers.0.self_attn.rotary_emb
Layer: model.layers.0.mlp
	Weight: gate_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: up_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: down_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.0.mlp.gate_proj
	Weight: weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.0.mlp.up_proj
	Weight: weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.0.mlp.down_proj
	Weight: weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.0.mlp.act_fn
Layer: model.layers.0.input_layernorm
	Weight: weight | Size: torch.Size([4096]) | Trainable: True
Layer: model.layers.0.post_attention_layernorm
	Weight: weight | Size: torch.Size([4096]) | Trainable: True
...
Layer: model.layers.31
	Weight: self_attn.q_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: self_attn.k_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: self_attn.v_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: self_attn.o_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: mlp.gate_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: mlp.up_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: mlp.down_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: input_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
	Weight: post_attention_layernorm.weight | Size: torch.Size([4096]) | Trainable: True
Layer: model.layers.31.self_attn
	Weight: q_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
	Weight: k_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: v_proj.weight | Size: torch.Size([2097152, 1]) | Trainable: False
	Weight: o_proj.weight | Size: torch.Size([8388608, 1]) | Trainable: False
Layer: model.layers.31.self_attn.q_proj
	Weight: weight | Size: torch.Size([8388608, 1]) | Trainable: False
Layer: model.layers.31.self_attn.k_proj
	Weight: weight | Size: torch.Size([2097152, 1]) | Trainable: False
Layer: model.layers.31.self_attn.v_proj
	Weight: weight | Size: torch.Size([2097152, 1]) | Trainable: False
Layer: model.layers.31.self_attn.o_proj
	Weight: weight | Size: torch.Size([8388608, 1]) | Trainable: False
Layer: model.layers.31.self_attn.rotary_emb
Layer: model.layers.31.mlp
	Weight: gate_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: up_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
	Weight: down_proj.weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.31.mlp.gate_proj
	Weight: weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.31.mlp.up_proj
	Weight: weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.31.mlp.down_proj
	Weight: weight | Size: torch.Size([29360128, 1]) | Trainable: False
Layer: model.layers.31.mlp.act_fn
Layer: model.layers.31.input_layernorm
	Weight: weight | Size: torch.Size([4096]) | Trainable: True
Layer: model.layers.31.post_attention_layernorm
	Weight: weight | Size: torch.Size([4096]) | Trainable: True
Layer: model.norm
	Weight: weight | Size: torch.Size([4096]) | Trainable: True
Layer: lm_head
	Weight: weight | Size: torch.Size([32000, 4096]) | Trainable: True
```

**Comment by BenjaminBossan** [comment]
> Maybe it worked because the data was very small. Still it is kind of very strange that it worked at all knowing this now

I'm not too surprised that it worked and loss improved on a small dataset. What would be surprising is if it worked on a larger scale. But there is also for instance [this paper](https://arxiv.org/abs/2312.11420) which only tunes the LN layers.

**Comment by alexcpn** [comment]
>But there is also for instance [this paper](https://arxiv.org/abs/2312.11420) which only tunes the LN layers.
This is cool! "... we intriguingly note that, within each attention block, tuning LayerNorm suffices to yield strong performance"

>What would be surprising is if it worked on a larger scale.
Is there any data set that you can suggest that I train on and test this further? Taking data from most sources on the internet, I find that the Mistral or LLAMa 13 billion parameter models have already been trained on. This is one reason I work with a synthetically created minimal data set.

**Comment by BenjaminBossan** [comment]
> Can you suggest any data set that I train on and test this further? Taking data from most sources on the internet, I find that the Mistral or LLAMa 13 billion parameter models have already been trained on. This is one reason I work with a synthetically created minimal data set.

That's indeed a big problem. You may have some success with instruction fine-tuning datasets, some have been applied to Llama, but they typically use full fine-tuning AFAIK. Would be interesting to compare the results.
